File: advent-of-code-rust/src/aoc2023/day1/trebuchet.rs
use regex::{Captures, Regex};

use crate::prelude::*;

pub struct Trebuchet;

const DIGITS_PATTERN: &'static str = r"^[[:alpha:]]*?(?<first>[0-9])[[:alnum:]]*?(?<second>[0-9])?[[:alpha:]]*?$";

impl solution::Solution for Trebuchet {
    fn solve_part_one(&self, input_content: &String) -> String {
        let Ok(re) = Regex::new(DIGITS_PATTERN) else { panic!("Invalid pattern {}", DIGITS_PATTERN) };

        input_content.lines()
            .map(|l| get_two_digit_number(l, &re))
            .sum::<i32>()
            .to_string()
    }

    fn solve_part_two(&self, input_content: &String) -> String {
        let forward_pattern = r"[0-9]|one|two|three|four|five|six|seven|eight|nine";
        let backward_pattern = r"[0-9]|enin|thgie|neves|xis|evif|ruof|eerht|owt|eno";

        let Ok(forward_re) = Regex::new(forward_pattern) else { panic!("Invalid pattern {}", forward_pattern) };
        let Ok(backward_re) = Regex::new(backward_pattern) else { panic!("Invalid pattern {}", backward_pattern) };

        input_content.lines()
            .map(|l| get_two_digit_number_with_bidirectional_search(l, &forward_re, &backward_re))
            .sum::<i32>()
            .to_string()
    }
}

fn get_two_digit_number(line: &str, re: &Regex) -> i32 {
    let Some(cap) = re.captures(line) else { panic!("Did not match regex: {}", line) };
    let (first_digit, second_digit) = extract_digits(&cap);
    let two_digit_string = first_digit.to_string() + second_digit;
    two_digit_string.parse().unwrap()
}

fn extract_digits<'a>(cap: &'a Captures) -> (&'a str, &'a str) {
    let first_digit = extract_digit(&cap["first"]);
    let second_digit = cap.name("second")
        .map_or(first_digit, |m| extract_digit(m.as_str()));
    (first_digit, second_digit)
}

fn get_two_digit_number_with_bidirectional_search(line: &str, forward_re: &Regex, backward_re: &Regex) -> i32 {
    let rev_line: String = line.chars().rev().collect();

    let first_match = forward_re.find(line).unwrap().as_str();
    let second_match_rev = backward_re.find(rev_line.as_str()).unwrap().as_str();

    let second_match: String = second_match_rev.chars().rev().collect();

    let first_digit = extract_digit(first_match);
    let second_digit = extract_digit(second_match.as_str());

    let two_digit_string = first_digit.to_string() + second_digit;
    two_digit_string.parse().unwrap()
}

fn extract_digit(matched: &str) -> &str {
    match matched {
        "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" => matched,
        "one" => "1",
        "two" => "2",
        "three" => "3",
        "four" => "4",
        "five" => "5",
        "six" => "6",
        "seven" => "7",
        "eight" => "8",
        "nine" => "9",
        _ => panic!("Could not parse {matched}")
    }
}

File: AdventOfCodePython/src/aoc2023/day1/trebuchet.py
import re

import src.aoc_template.solution as solution

VALID_DIGIT_WORDS = "one|two|three|four|five|six|seven|eight|nine"


def _solve(input_lines: list[str], get_calibration_value) -> str:
    calibration_values = map(lambda line : get_calibration_value(line), input_lines)
    return str(sum(calibration_values))


def _get_digit(text):
    if len(text) == 1: return text
    word_list = VALID_DIGIT_WORDS.split("|")
    value = word_list.index(text) + 1
    return str(value)

def _get_calibration_value_p1(line: str) -> int:
    characters = re.findall("[\d]", line)
    return int(characters[0] + characters[-1])

def _get_calibration_value_p2(line: str) -> int:
    characters = re.findall(f'(?=(\d|{VALID_DIGIT_WORDS}))', line)

    digit1 = _get_digit(characters[0])
    digit2 = _get_digit(characters[-1])
    return int(digit1 + digit2)

class Trebuchet(solution.AbstractSolution):
    def solve_part_one(self, input_lines: list[str]) -> str:
        return _solve(input_lines, _get_calibration_value_p1)

    def solve_part_two(self, input_lines: list[str]) -> str:
        return _solve(input_lines, _get_calibration_value_p2)
